loc-analysis.exs

```diff
@@ -0,0 +1,202 @@
+    end
+  end
+end
+
+defmodule Git do
+  def get_snapshot!(commit_hash) do
+    with destination_path <- System.tmp_dir!(),
+         {:ok, tar_file} <- call_git_archive!(commit_hash, destination_path),
+         :ok <- call_tar_extract!(tar_file, destination_path),
+         :ok <- Path.join(destination_path, tar_file) |> File.rm!() do
+      {:ok, Path.join(destination_path, commit_hash)}
+    end
+  end
+
+  def monthly_commits do
+    call_git_log!()
+    |> String.split("\n")
+    |> Stream.uniq_by(&year_month_of/1)
+  end
+
+  def weekly_commits do
+    call_git_log!()
+    |> String.split("\n")
+    |> Stream.uniq_by(&beginning_of_week/1)
+  end
+
+  def call_git_log! do
+    case System.cmd("git", [
+           "log",
+           "--pretty=format:%ad,%H",
+           "--date=iso",
+           "--date=short",
+           "--reverse"
+         ]) do
+      {log, 0} -> log
+      {error_msg, _} -> raise GitCommandError, error_msg
+    end
+  end
+
+  def parse_git_log_line(log_line), do: log_line |> String.split(",")
+
+  defp year_month_of(commit), do: commit |> String.slice(0..6)
+
+  defp beginning_of_week(commit),
+    do:
+      commit
+      |> String.slice(0..9)
+      |> Date.from_iso8601()
+      |> elem(1)
+      |> Date.beginning_of_week()
+      |> Date.to_string()
+
+  defp call_git_archive!(commit_hash, temp_dir_path_str) do
+    System.cmd("git", [
+      "archive",
+      commit_hash,
+      "--prefix=#{commit_hash}/",
+      "--format=tar",
+      "--output=#{temp_dir_path_str}/#{commit_hash}.tar"
+    ])
+    |> case do
+      {_, 0} -> {:ok, "#{commit_hash}.tar"}
+    end
+  end
+
+  defp call_tar_extract!(archive_file, archive_location) do
+    System.cmd(
+      "tar",
+      [
+        "xf",
+        archive_file
+      ],
+      cd: archive_location
+    )
+    |> case do
+      {_, 0} -> :ok
+    end
+  end
+end
+
+defmodule Analysis do
+  def monthly(file) do
+    Git.monthly_commits()
+    |> Task.async_stream(&run_loc_analysis!/1, on_timeout: :kill_task, timeout: 30_000)
+    |> Stream.reject(fn task_result -> {:exit, _} |> match?(task_result) end)
+    |> Stream.map(fn {:ok, analysis} -> analysis end)
+    |> Enum.to_list()
+    |> Enum.map_reduce(
+      %MapSet{},
+      fn analysis, headers ->
+        {
+          analysis,
+          analysis |> Map.keys() |> MapSet.new() |> MapSet.union(headers)
+        }
+      end
+    )
+    |> then(fn {analysis_arr, headers} ->
+      analysis_arr |> CSV.encode(headers: Enum.to_list(headers))
+    end)
+    |> Stream.each(&IO.write(file, &1))
+    |> Enum.to_list()
+  end
+
+  def weekly(file) do
+    Git.weekly_commits()
+    |> Task.async_stream(&run_loc_analysis!/1, on_timeout: :kill_task, timeout: 30_000)
+    |> Stream.reject(fn task_result -> {:exit, _} |> match?(task_result) end)
+    |> Stream.map(fn {:ok, analysis} -> analysis end)
+    |> Enum.to_list()
+    |> Enum.map_reduce(
+      %MapSet{},
+      fn analysis, headers ->
```

Comment:
My [comment above](https://github.com/brkn/code-churn/pull/1/files/4892b1437d017c847c62b76e222c4301fa5cacac#r1318551903) feels even more relevant now that this function does the same :)

Reply:
definetely.

Would you suggest merging two methods into single method and pass the commit list as a parameter?

```ex
def generate_analysis_csv(commit_list, file) do
	commit_list
	|> Task.async_stream(&run_loc_analysis!/1, on_timeout: :kill_task, timeout: 30_000)
	|> Stream.reject(fn task_result -> {:exit, _} |> match?(task_result) end)
	|> Stream.map(fn {:ok, analysis} -> analysis end)
	|> Enum.to_list()
	|> Enum.map_reduce(&aggregate_headers/1)
	|> then(fn {analysis_arr, headers} ->
      analysis_arr |> CSV.encode(headers: Enum.to_list(headers))
    end)
    |> Stream.each(&IO.write(file, &1))
    |> Enum.to_list()
end
```

Reply:
Yes, that was roughly my idea!

Another one: you may have read that Streams are lazy, while Enums are eager. The moment you use `Enum.map_reduce`, you lose the laziness and start traversing the whole list at  every stage in the pipeline.

Do you think there's a way of swapping map_reduce with, uh, `Stream.resource/3` or `Stream.unfold/2`? It's an interesting exercise.